cvat/apps/engine/views.py

```diff
@@ -851,16 +859,179 @@ def jobs(self, request, pk):
+            # file uploading can be restarted. In such case we'll
+            # just reuse the first position of this file
+            pass
+
+    class _InvalidMetafileError(Exception):
+        """
+        Indicates an invalid upload metafile found
+        """
+
+    def _sort_uploaded_files(self, uploaded_files: List[Dict[str, Any]]) -> List[str]:
+        """
+        Restores file ordering for the "predefined" file sorting method of the task creation.
+        Without this, the file order is defined by os.listdir(), which returns files unordered.
+        Read more: https://github.com/opencv/cvat/issues/5061
+        """
+
+        upload_dir = self. get_upload_dir()
+
+        upload_metafile = None
+        for filename in self._get_tus_metafile_names():
+            if osp.isfile(osp.join(upload_dir, filename)):
+                upload_metafile = filename
+                break
+
+        if not upload_metafile:
+            raise FileNotFoundError("Uploaded file order is not specified")
+
+        expected_files = self._read_tus_upload_meta_file(osp.join(upload_dir, upload_metafile))
+        expected_files.append(upload_metafile)
+
+        uploaded_file_names = { f: f for f in uploaded_files }
+        mismatching_files = list(uploaded_file_names.keys() ^ expected_files)
+        if mismatching_files:
+            DISPLAY_ENTRIES_COUNT = 5
+            mismatching_display = [
+                fn + (" (extra)" if fn in uploaded_file_names else " (missing)")
+                for fn in mismatching_files[:DISPLAY_ENTRIES_COUNT]
+            ]
+            remaining_count = len(mismatching_files) - DISPLAY_ENTRIES_COUNT
+            raise self._InvalidMetafileError(
+                "Uploaded files do no match the upload file list contents. "
+                "Please check the upload metainfo and the list of uploaded files. "
+                "Mismatching files: {}{}"
+                .format(
+                    ", ".join(mismatching_display),
+                    f" (and {remaining_count} more). " if 0 < remaining_count else ""
+                )
+            )
+
+        return [uploaded_file_names[fn] for fn in expected_files]
+
+    # UploadMixin method
+    def init_file_upload(self, request):
+        response = super().init_file_upload(request)
+
+        if self._is_data_uploading() and response.status_code == status.HTTP_201_CREATED:
+            metafile = osp.join(self.get_upload_dir(), self._get_tus_input_ordering_metafile_name())
+            if osp.isfile(metafile):
+                self._maybe_append_tus_upload_metafile_entry(metafile, response['Upload-Filename'])
+
+        return response
+
+    # UploadMixin method
+    def append_files(self, request):
+        client_files = self._get_request_client_files(request)
+        if self._is_data_uploading() and client_files:
+            metafile = osp.join(self.get_upload_dir(), self._get_tus_input_ordering_metafile_name())
```

Comment:
As I understand, the client files are unordered - but then why would we record their order in the manifest? It's going to be meaningless. I think this method of uploading should just be disallowed when the sorting order is predefined and there was no initial file list.

Reply:
Client files are always ordered, but in some cases the order is different from what is needed. Because of this there are 2 modes of uploading with TUS: with the input order and with the custom order specified explicitly. By default, the input order is used, so there are no extra actions needed from the user to utilize the predefined sorting method. If the user wishes to send files unordered and then sort them on the server side - it will require sending the desired filename list in a special request (the Upload-Start).